distsql: fix non-consecutive span planner crash

RaduBerinde opened the pull request:
The planning code was making some incorrect assumptions about `SpanResolverIterator`, specifically that the ranges returned are non-overlapping (which can happen because caches are updated asynchronously). Fixing the code to tolerate overlapping ranges.

Adding a test that reproduces the panic in #15249.

Fixes #15249.

The second commit fixes another problem uncovered by stress-running the new test.

RaduBerinde committed:
distsql: fix non-consecutive span planner crash

The planning code was making some incorrect assumptions about
`SpanResolverIterator`, specifically that the ranges returned are
non-overlapping (which can happen because caches are updated asynchronously).
Fixing the code to tolerate overlapping ranges.

Adding a test that reproduces the panic in #15249.

Fixes #15249.

RaduBerinde committed:
testcluster: fix SplitRange

The `SplitRange` function has a problematic corner case: if the ranges are
already split at the particular split point, what it reads as the left
descriptor is actually the right descriptor. This descriptor can have
`EndKey=/Max` and looking up the descriptor starting at that key loops forever
in `RangeIterator.Seek()`.

The code has an error case for this but it is ineffective because the split
point is not exactly equal to the split key (in fact, it's always a prefix).

Fixing by just scanning and reverse-scanning for meta ranges around the split
key.

Also adding an error check in `RangeIterator.Seek` when the starting key is not
valid.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15260)
<!-- Reviewable:end -->

tamird commented:
Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/server/testserver.go, line 520 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiHbOzjGEWWrigeIRFD:-KiHbOzjGEWWrigeIRFE:b-ndptff) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5751185a3f6803000c45c39068fa086836ca32d/pkg/server/testserver.go#L520)):*
> ```Go
> 	// We use a transaction so that we get consistent results between the two
> 	// scans (in case there are other splits happening).
> 	txn := client.NewTxn(ts.DB())
> ```

why not use the closure form? the transaction doesn't outlive this function.

---

*[pkg/sql/distsql_physical_planner.go, line 408 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiH_qB5wldugPRVi_GP:-KiH_qB5wldugPRVi_GQ:b-qvlyxy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4262717916eba310111fcee86bfcfcf56faf153d/pkg/sql/distsql_physical_planner.go#L408)):*
> ```Go
> 
> 		var lastNodeID roachpb.NodeID
> 		// lastKey maintains the EndKey of the last span piece.
> ```

s/piece//, i think

---

*[pkg/sql/distsql_physical_planner.go, line 420 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiHaFYDGfTS3V52fzqb:-KiHaFYDGfTS3V52fzqc:bs9p637) ([raw file](https://github.com/cockroachdb/cockroach/blob/4262717916eba310111fcee86bfcfcf56faf153d/pkg/sql/distsql_physical_planner.go#L420)):*
> ```Go
> 			desc := it.Desc()
> 
> 			if lastKey.Less(desc.StartKey) || !lastKey.Less(desc.EndKey) {
> ```

Can you use `desc.ContainsKey`?

---

*[pkg/sql/distsql_physical_planner.go, line 423 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiHa3FgR6ussvfXNDfr:-KiHa3FgR6ussvfXNDfs:bhiywhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/4262717916eba310111fcee86bfcfcf56faf153d/pkg/sql/distsql_physical_planner.go#L423)):*
> ```Go
> 				// This range must contain the last range's EndKey.
> 				log.Fatalf(
> 					ctx, "next range doesn't cover last end key: %#v %v",
> ```

don't you want to include `lastKey` here?

---

*[pkg/sql/distsql_physical_planner_test.go, line 128 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiH_1Q4UTeFYepoDfKH:-KiH_1Q4UTeFYepoDfKI:bb7be76) ([raw file](https://github.com/cockroachdb/cockroach/blob/4262717916eba310111fcee86bfcfcf56faf153d/pkg/sql/distsql_physical_planner_test.go#L128)):*
> ```Go
> 
> 				splitKey := keys.MakeRowSentinelKey(pik)
> 				if _, _, err = tc.Server(0).SplitRange(splitKey); err != nil {
> ```

`:=`

---

*[pkg/sql/distsql_physical_planner_test.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiH_J2buhnVMzj4dPVt:-KiH_J2buhnVMzj4dPVu:b4hp0wk) ([raw file](https://github.com/cockroachdb/cockroach/blob/4262717916eba310111fcee86bfcfcf56faf153d/pkg/sql/distsql_physical_planner_test.go#L163)):*
> ```Go
> 			if err != nil {
> 				t.Error(err)
> 				return
> ```

you can avoid all these returns by wrapping in a function that returns an error. optional.

---

*[pkg/sql/distsql_physical_planner_test.go, line 167 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiH_GzSMozrgaYiFbkO:-KiH_GzSMozrgaYiFbkP:b-5wok37) ([raw file](https://github.com/cockroachdb/cockroach/blob/4262717916eba310111fcee86bfcfcf56faf153d/pkg/sql/distsql_physical_planner_test.go#L167)):*
> ```Go
> 
> 			defer func() {
> 				_ = goDB.Close()
> ```

may as well check the error, no?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: remove leftover SQLRunner.Subtest

RaduBerinde commented:
Review status: 1 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/server/testserver.go, line 520 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiHbOzjGEWWrigeIRFD:-KiIGNTGs8Ek6GAVmVCF:boopawo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5751185a3f6803000c45c39068fa086836ca32d/pkg/server/testserver.go#L520)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why not use the closure form? the transaction doesn't outlive this function.
</blockquote></details>

TFTR! Done.

---

*[pkg/sql/distsql_physical_planner.go, line 408 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiH_qB5wldugPRVi_GP:-KiIGWzk8k1-qS1y99Jb:b-fa9sz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4262717916eba310111fcee86bfcfcf56faf153d/pkg/sql/distsql_physical_planner.go#L408)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/piece//, i think
</blockquote></details>

clarified

---

*[pkg/sql/distsql_physical_planner.go, line 420 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiHaFYDGfTS3V52fzqb:-KiIGqnbGhaIlFAdIE5I:btojbfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/4262717916eba310111fcee86bfcfcf56faf153d/pkg/sql/distsql_physical_planner.go#L420)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Can you use `desc.ContainsKey`?
</blockquote></details>

Great!

---

*[pkg/sql/distsql_physical_planner.go, line 423 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiHa3FgR6ussvfXNDfr:-KiIGuKCJuTt_u3bU0OI:b-nuqn3t) ([raw file](https://github.com/cockroachdb/cockroach/blob/4262717916eba310111fcee86bfcfcf56faf153d/pkg/sql/distsql_physical_planner.go#L423)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

don't you want to include `lastKey` here?
</blockquote></details>

it's part of the last entry of `splits`

---

*[pkg/sql/distsql_physical_planner_test.go, line 128 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiH_1Q4UTeFYepoDfKH:-KiIH3APwz7eSKvrLeu3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4262717916eba310111fcee86bfcfcf56faf153d/pkg/sql/distsql_physical_planner_test.go#L128)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`:=`
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner_test.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiH_J2buhnVMzj4dPVt:-KiIHFKIJtUEVpCF70OZ:b-1aq1iw) ([raw file](https://github.com/cockroachdb/cockroach/blob/4262717916eba310111fcee86bfcfcf56faf153d/pkg/sql/distsql_physical_planner_test.go#L163)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you can avoid all these returns by wrapping in a function that returns an error. optional.
</blockquote></details>

I started that way but I find it harder to read overall (given this is all already inside of a func)

---

*[pkg/sql/distsql_physical_planner_test.go, line 167 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiH_GzSMozrgaYiFbkO:-KiIHYIepMoUr5be6hAz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4262717916eba310111fcee86bfcfcf56faf153d/pkg/sql/distsql_physical_planner_test.go#L167)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

may as well check the error, no?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r4, 2 of 2 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/server/testserver.go, line 555 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiMeu6uQWhpwfRxPTpU:-KiMeu6vzuWkQuSFpXEk:bmj1eql) ([raw file](https://github.com/cockroachdb/cockroach/blob/80a6d4a501e6f55322284c381779cbe8f2cb6131/pkg/server/testserver.go#L555)):*
> ```Go
> 
> 		if !leftRangeDesc.EndKey.Equal(rightRangeDesc.StartKey) {
> 			return fmt.Errorf(
> ```

errors.Errorf, since you're already importing errors [optional]

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 2 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/server/testserver.go, line 555 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-KiMeu6uQWhpwfRxPTpU:-KiPIlk7nYYIbvquVDYF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/80a6d4a501e6f55322284c381779cbe8f2cb6131/pkg/server/testserver.go#L555)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

errors.Errorf, since you're already importing errors [optional]
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r7, 2 of 2 files at r8, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15260#-:-KiPZLmmm5cY6CdgNAEC:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
distsql: fix memory monitor race condition

Fixing a race condition in the distsqlrun server, where we were setting the
monitor in the server's `evalCtx` unnecessarily. The race is innocuous but fails
`TestPlannerDuringSplits` under `stressrace`.

RaduBerinde commented:
CC @justinj for the fourth commit.

justinj commented:
Good call!
